pysmt/cnf.py

```diff
@@ -0,0 +1,217 @@
+#
+# This file is part of pySMT.
+#
+#   Copyright 2014 Andrea Micheli and Marco Gario
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+import pysmt.shortcuts as shortcuts
+from pysmt.walkers import DagWalker
+from pysmt.typing import BOOL
+
+class CNFizer(DagWalker):
+
+    THEORY_PLACEHOLDER = "__Placeholder__"
+
+    TRUE_CNF = frozenset()
+    FALSE_CNF = frozenset([frozenset()])
+
+    def __init__(self, environment=None):
+        DagWalker.__init__(self)
+        self.env = environment if environment else shortcuts.get_env()
+        self.mgr = self.env.formula_manager
+        self._introduced_variables = {}
+        self._cnf_pieces = {}
+
+    def _key_var(self, formula):
+        if formula in self._introduced_variables:
+            res = self._introduced_variables[formula]
+        else:
+            res = self.mgr.FreshSymbol()
+            self._introduced_variables[formula] = res
+        return res
+
+    def convert(self, formula):
+        tl, cnf = self.walk(formula)
+        res = [frozenset([tl])]
+        for clause in cnf:
+            if len(clause) == 0:
+                return CNFizer.FALSE_CNF
+            simp = []
+            for lit in clause:
+                if lit.is_true():
+                    simp = None
+                    break
+                elif not lit.is_false():
+                    simp.append(lit)
+            if simp:
+                res.append(frozenset(simp))
+        return frozenset(res)
+
+    def convert_as_formula(self, formula):
+        lsts = self.convert(formula)
+
+        conj = []
+        for clause in lsts:
+            if len(clause) == 0:
+                return self.mgr.FALSE()
+            elif len(clause) == 1:
+                conj.append(next(iter(clause)))
+            else:
+                conj.append(self.mgr.Or(clause))
+
+        if len(conj) == 0:
```

Comment:
This should be done automatically by And. and of an empty list is TRUE, of a singleton is the elemtn.